src/midiEditorCommands.cpp

```diff
@@ -142,27 +142,30 @@ PreviewSource previewSource;
 preview_register_t previewReg = {0};
 
 // Queue note off events for the notes currently being previewed.
-// This function doesn't begin sending the events.
-void previewNotesOff() {
+// when sendNoteOff is true, this function  also sends the events.
+void previewNotesOff(bool sendNoteOff) {
 	for (auto note = previewingNotes.cbegin(); note != previewingNotes.cend(); ++note) {
 		MIDI_event_t event = {0, 3, {
 			(unsigned char)(MIDI_NOTE_OFF | note->channel),
 			(unsigned char)note->pitch, (unsigned char)note->velocity}};
 		previewSource.events.push_back(event);
 	}
+	if (sendNoteOff) {
+		// Send the events.
+		previewReg.curpos = 0.0;
+		PlayTrackPreview(&previewReg);
+	}
+	previewingNotes.clear();
 }
 
 // Called after the preview length elapses to turn off notes currently being previewed.
 void CALLBACK previewDone(HWND hwnd, UINT msg, UINT_PTR event, DWORD time) {
-	if (event != previewDoneTimer)
+	if (event != previewDoneTimer) {
 		return; // Cancelled.
-	previewNotesOff();
-	previewingNotes.clear();
-	// Send the events.
-	previewReg.curpos = 0.0;
-	PlayTrackPreview(&previewReg);
-	previewDoneTimer = 0;
-	KillTimer(nullptr, event);
+	}
+	if (cancelMidiPreviewDoneTimer()) {
```

Comment:
I assume there should never be a case where cancelMidiPreviewDoneTimer returns false here, since we already returned early if the timer was cancelled? I guess there's nothing wrong with checking the result anyway, but thought I'd check.

Reply:
I wonder whether it could be theoretically possible that the preview is canceled at the same time previewDone is called. In that case, previewDoneTimer would be nullptr when the event occurs.

Reply:
It's entirely possible for the callback to still run after the timer is cancelled, yes. However, we set previewDoneTimer to nullptr in that case, and the early return above this (`event != previewDoneTimer`) will happen if previewDoneTimer is nullptr (or a subsequently set timer).